src/databricks/labs/ucx/hive_metastore/tables.py

```diff
@@ -481,3 +481,75 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
         except Exception as e:  # pylint: disable=broad-exception-caught
             logger.error(f"Couldn't fetch information for table {full_name} : {e}")
             return None
+
+
+class FasterTableScanHack:
+    def __init__(self, spark: any):
+        self._spark = spark
+
+    @cached_property
+    def _external_catalog(self):
+        return self._spark._jsparkSession.sharedState().externalCatalog()
+
+    def _iterator(self, result: any):
+        it = result.iterator()
+        while it.hasNext():
+            yield it.next()
+
+    def list_tables(self, database: str) -> list[str]:
+        try:
+            return list(self._iterator(self._external_catalog.listTables(database)))
+        except Exception as err:
+            logger.warning(f"Failed to list tables in {database}: {err}")
+
+    def get_table(self, database: str, table: str) -> any:
+        raw_table = self._external_catalog.getTable(database, table)
+        table_format = raw_table.provider().getOrElse(None)
+        if not table_format:
+            table_format = "UNKNOWN"
+        location_uri = raw_table.storage().locationUri().getOrElse(None)
+        if location_uri:
+            location_uri = location_uri.toString()
+        view_text = raw_table.viewText()
+        return Table(
+            catalog='hive_metastore',
+            database=database,
+            name=table,
+            object_type=raw_table.tableType().name(),
+            table_format=table_format,
+            location=location_uri,
+            view_text=view_text,
+        )
+
+    def _all_databases(self) -> list[str]:
+        if not self._include_database:
+            return [row[0] for row in self._fetch("SHOW DATABASES")]
```

Comment:
there's no fetch method here. and use the previous optimized code to fetch databases. handle exceptions.